File: newsource.pl
#!/usr/bin/env perl

sub openNamespace {
	print FILE  "namespace @_[1] {\n";
}

sub closeNamespace {
	print FILE "} // namespace @_[1]\n";
}

sub classDefinition {
	print FILE "class @_[1] {\n";
	print FILE "public:\n";
	print FILE "\t@_[1]();\n";
	print FILE "\t~@_[1]();\n";
	print FILE "};\n";
}

sub printCopyright {
	# Copyright notice . . .
	print FILE <<EOF;
/**
	Aesalon, a tool to visualize a program's behaviour at run-time.
	Copyright (C) 2010, Aesalon Development Team.

	Aesalon is distributed under the terms of the GNU GPLv3. For more
	licensing information, see the file LICENSE included with the distribution.
	
	\@file @_[0]
*/

EOF
}

sub createHeader {
	if(@_[1] ne "") {
		$filename = @_[0] . (lc @_[1]) . "/" . @_[2] . ".h\n";
	}
	else {
		$filename = @_[0] . "/" . @_[2] . ".h\n";
	}

	open(FILE, ">$filename") or die("Can't open $filename . . .");

	printCopyright($filename);

	# Inclusion guard . . .
	#$guard = "Aesalon" . @_[3] . "_" . @_[1] . "_" . @_[2] . "_H";
	if(@_[1] ne "") {
		$guard = "Aesalon@_[3]_@_[1]_@_[2]_H";
	}
	else {
		$guard = "Aesalon@_[3]_@_[2]_H";
	}
	
	# Remove duplicate . . .
	#$guard =~ s/@_[1]@_[1]/@_[1]/e;
	
	print FILE "#ifndef $guard\n";
	print FILE "#define $guard\n\n";

	openNamespace(FILE, @_[3]);
	openNamespace(FILE, @_[1]) if @_[1] ne "";

	print FILE "\n";

	classDefinition(FILE, @_[2]);

	print FILE "\n";

	closeNamespace(FILE, @_[1]) if @_[1] ne "";
	closeNamespace(FILE, @_[3]);

	print FILE "\n#endif\n";

	close(FILE);
}

sub createSource {
	$filename = @_[0] . (lc @_[1]) . "/" . @_[2] . ".cpp\n";

	open(FILE, ">$filename") or die("Can't open $filename . . .");
	
	printCopyright($filename);
	
	if(@_[1] ne "") {
		print FILE "#include \"" . (lc @_[1]) . "/@_[2].h\"\n\n";
	}
	else {
		print FILE "#include \"@_[2].h\"\n\n";
	}
	
	openNamespace(FILE, @_[3]);
	openNamespace(FILE, @_[1]) if @_[1] ne "";

	print FILE "\n\n\n";

	closeNamespace(FILE, @_[1]) if @_[1] ne "";
	closeNamespace(FILE, @_[3]);

	close(FILE);
}

sub extractNamespace {
	if(!(@_[0] =~ m/(.+)\:\:(.+)/)) { return "", @_[0]; }
	
#	print "1: $1, 2: $2\n";
	
	if($2 eq "") { return "", $1; }
	return $1, $2
}

$argumentCount = $#ARGV + 1;

if($argumentCount < 1) {
	die("Not enough arguments given");
}

$mode = "";

# Parse mode . . .
if($argumentCount == 1) {
	$mode = "common";
}
else {
	if(lc $ARGV[0] eq "monitor") { $mode = "monitor"; }
	elsif(lc $ARGV[0] eq "visualizer") { $mode = "visualizer"; }
	elsif(lc $ARGV[0] eq "common") { $mode = "common"; }
	else {
		die "Unknown mode";
	}
}

if($mode eq "monitor") {
	createHeader("include/monitor/", extractNamespace($ARGV[1]), "Monitor");
	createSource("monitor/src/", extractNamespace($ARGV[1]), "Monitor");
}
elsif($mode eq "visualizer") {

}
elsif($mode eq "common") {
	createHeader("include/common", extractNamespace($ARGV[1]), "Common");
}

File: newsource.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import os
import string

config = {
	"project":		"Aesalon",
	"desc":			"a tool to visualize program behaviour in real time",
	"authors":		"Aesalon development team",
	"license":		"GNU GPLv3",
	"devyears":		"2009-2011",
	"website":		"http://aesalon.berlios.de/",
	"srcBase":		"src",
	"incBase":		"include",
	"guardBase":	"Aesalon",
	"fileHeader":
"""/** %(project)s, %(desc)s.
	Copyright (C) %(devyears)s, %(authors)s.
	
	%(project)s is distributed under the terms of the %(license)s. See
	the included file LICENSE for more information.
	
	@file %(path)s
*/
""",
	"cHeader":
"""
#ifndef %(includeGuard)s
#define %(includeGuard)s



#endif
""",

	"cSource":
"""
#include "%(incPath)s"


""",
	"cppHeader":
"""
#ifndef %(includeGuard)s
#define %(includeGuard)s

""",
	"openNamespace":
"""
namespace %(namespace)s {
""",
	
	"cppClassDef":
"""
class %(className)s {
public:
	%(className)s();
	~%(className)s();
};
""",
	"closeNamespace":
"""
} // namespace %(namespace)s
""",
	"cppHeaderEnd":
"""
#endif

"""
}

class Generator:
	def __init__(self):
		self.__name = "Unknown name"
		pass

	def name(self):
		return self.__name
	def setName(self, name):
		self.__name = name

	def generate(self, name):
		pass

class CHeader(Generator):
	def __init__(self):
		Generator.__init__(self)
		self.setName("cheader")
	
	def generate(self, name):
		path = name
		path += ".h"
		
		fp = file(os.path.join(config["incBase"], path), "w")
		
		includeGuard = config["guardBase"] + " " + name.replace("/", " ")
		while True:
			index = includeGuard.find(" ")
			if index == -1: break
			
			includeGuard = includeGuard[:index] + includeGuard[index+1].upper() + includeGuard[index+2:]
		
		includeGuard += "_H"
		
		fileConfig = config
		fileConfig["path"] = path
		fileConfig["includeGuard"] = includeGuard
		
		fp.write(config["fileHeader"] % fileConfig)
		fp.write(config["cHeader"] % fileConfig)
		
		fp.close()
		
		return path

class CSource(Generator):
	def __init__(self):
		Generator.__init__(self)
		self.setName("csource")
	
	def generate(self, name, incPath=""):
		path = os.path.join(config["srcBase"], name)
		path += ".c"
		
		fp = file(path, "w")
		
		fileConfig = config
		fileConfig["path"] = path
		fileConfig["incPath"] = incPath
		
		fp.write(fileConfig["fileHeader"] % fileConfig)
		if incPath != "":
			fp.write(fileConfig["cSource"] % fileConfig)
		
		fp.close()

class CGenerator(Generator):
	def __init__(self):
		Generator.__init__(self)
		self.setName("c")
	
	def generate(self, name):
		CHeader().generate(name)
		CSource().generate(name, True)



class CppHeader(Generator):
	def __init__(self):
		Generator.__init__(self)
		self.setName("cppheader")
	
	def generate(self, name):
		path = name
		path = path.replace("::", "/")
		last = path.rfind("/")
		if last != -1:
			path = path[:last].lower() + path[last:]
		path += ".h"
		filepath = os.path.join(config["incBase"], path)
		
		fp = file(filepath, "w")
		
		includeGuard = config["guardBase"] + " " + name.replace("/", "_")
		includeGuard = includeGuard.replace("::", "_")
		while True:
			index = includeGuard.find(" ")
			if index == -1: break
			
			includeGuard = includeGuard[:index] + includeGuard[index+1].upper() + includeGuard[index+2:]
		
		includeGuard += "_H"
		
		fileConfig = config
		fileConfig["path"] = filepath
		fileConfig["includeGuard"] = includeGuard
		
		fp.write(config["fileHeader"] % fileConfig)
		fp.write(config["cppHeader"] % fileConfig)
		
		namespaceList = name.split("::")
		if len(namespaceList) != 1:
			className = namespaceList[-1]
			namespaceList = namespaceList[:-1]
			for namespace in namespaceList:
				fileConfig["namespace"] = namespace
				fp.write(fileConfig["openNamespace"] % fileConfig)
			
			fileConfig["className"] = className
			fp.write(fileConfig["cppClassDef"] % fileConfig)
			
			for namespace in namespaceList[::-1]:
				fileConfig["namespace"] = namespace
				fp.write(fileConfig["closeNamespace"] % fileConfig)
		
		fp.write(config["cppHeaderEnd"] % fileConfig)
		
		fp.close()
		
		return path

class CppSource(Generator):
	def __init__(self):
		Generator.__init__(self)
		self.setName("cppsource")
	
	def generate(self, name, incPath=""):
		path = os.path.join(config["srcBase"], name)
		path = path.replace("::", "/")
		last = path.rfind("/")
		if last != -1:
			path = path[:last].lower() + path[last:]
		path += ".cpp"
		
		fp = file(path, "w")
		
		fileConfig = config
		fileConfig["path"] = path
		fileConfig["incPath"] = incPath
		
		fp.write(fileConfig["fileHeader"] % fileConfig)
		if incPath != "":
			fp.write(fileConfig["cSource"] % fileConfig)
		
		namespaceList = name.split("::")
		if len(namespaceList) != 1:
			className = namespaceList[-1]
			namespaceList = namespaceList[:-1]
			for namespace in namespaceList:
				fileConfig["namespace"] = namespace
				fp.write(fileConfig["openNamespace"] % fileConfig)
			
			fp.write("\n\n")
			
			for namespace in namespaceList[::-1]:
				fileConfig["namespace"] = namespace
				fp.write(fileConfig["closeNamespace"] % fileConfig)
		
		fp.close()

class CppGenerator(Generator):
	def __init__(self):
		Generator.__init__(self)
		self.setName("cpp")
	
	def generate(self, name):
		incPath = CppHeader().generate(name)
		CppSource().generate(name, incPath)



def main():
	generatorMap = dict()
	def addGenerator(generator):
		generatorMap[generator.name()] = generator
	
	addGenerator(CHeader())
	addGenerator(CSource())
	addGenerator(CGenerator())
	
	addGenerator(CppHeader())
	addGenerator(CppSource())
	addGenerator(CppGenerator())
	
	if len(sys.argv) < 3:
		print "Usage:", sys.argv[0], "generator name1 [name2 name3 . . .]"
		print "Where generator is one of:"
		for g in generatorMap.values():
			print "\t", g.name()
		return
	generator = generatorMap[sys.argv[1]]
	
	for name in sys.argv[2:]:
		generator.generate(name)

main()
